### PR TITLE
Deny unsafe operations in unsafe functions in libstd/alloc.rs

### PR DESCRIPTION
**Pull request by poliorcetics**
Partial fix of #73904.

This encloses `unsafe` operations in `unsafe fn` in `libstd/alloc.rs`.

@rustbot modify labels: F-unsafe-block-in-unsafe-fn

### CODE REVIEW

File: src/libstd/alloc.rs [hunk 1]
```diff
@@ -300,13 +337,19 @@ pub mod __default_lib_allocator {
 
     #[rustc_std_internal_symbol]
     pub unsafe extern "C" fn __rdl_alloc(size: usize, align: usize) -> *mut u8 {
-        let layout = Layout::from_size_align_unchecked(size, align);
-        System.alloc(layout)
+        // SAFETY: see the guarantees expected by `Layout::from_size_align` and
+        // `GlobalAlloc::alloc`.
```
Comment (LukasKalbertodt) [inline src/libstd/alloc.rs:341]: But is it documented somewhere that the caller of `__rdl_alloc` needs to make sure the parameters are valid according to `Layout::from_size_align`? Same for the functions below.

**Review by LukasKalbertodt — APPROVED** [review]
Thanks! I left one comment, but that's rather a question. This PR is certainly an improvement.

File: src/libstd/alloc.rs (hunk 1 above)
Comment (poliorcetics) [reply src/libstd/alloc.rs:341]: Not to my knowledge, or at least I do not remember finding anything when searching for it. I read the code several times to check and saw nothing that invalidated that but I may have missed something.

**Pull request merged**